main/src/ca/uwaterloo/flix/language/phase/EntryPoints.scala

```diff
@@ -473,6 +474,129 @@ object EntryPoints {
+      * }
+      * }}}
+      *
+      * Invalid handler example
+      * {{{
+      * @DefaultHandler
+      * def handle(f: Unit -> a \ ef, arg: tpe): Bool \ (ef - ef2) + PrimitiveEffect1 + ... = exp
+      * }}}
+      *
+      * @param handlerSym The symbol of the handler function
+      * @param handlerDef The definition of the handler function
+      * @param root       The typed AST root
+      * @return Either a valid HandlerInfo or a list of entry point errors
+      */
+    private def checkHandler(handlerSym: Symbol.DefnSym, handlerDef: TypedAst.Def, root: TypedAst.Root)(implicit flix: Flix): Validation[HandlerInfo, EntryPointError] = {
+      // First we obtain the effect corresponding to the handler
+      val effFqn = handlerSym.namespace.mkString(".")
+      val effSymbol = Symbol.mkEffSym(effFqn)
+      val companionEffect = root.effects.get(effSymbol)
+      companionEffect match {
+        // There is a valid effect to wrap
+        case Some(_) =>
+          // Check that handler has a signature like
+          // def handle(f: Unit -> a \ ef): a \ (ef - efHandle) + PrimitiveEffect1 + ...
+          val efHandle = Type.Cst(TypeConstructor.Effect(effSymbol, Kind.Eff), SourceLocation.Unknown)
+          val a = Type.freshVar(Kind.Star, SourceLocation.Unknown)
+          val ef = Type.freshVar(Kind.Eff, SourceLocation.Unknown)
+          // Create (IO & opt1) to make it optional
+          val opt1 = Type.freshVar(Kind.Eff, SourceLocation.Unknown)
+          val optionalIO = Type.mkIntersection(Type.IO, opt1, SourceLocation.Unknown)
+          // Create (NonDet & opt2) to make it optional
+          val opt2 = Type.freshVar(Kind.Eff, SourceLocation.Unknown)
+          val optionalNonDet = Type.mkIntersection(Type.NonDet, opt2, SourceLocation.Unknown)
+          val schemeToCheckAgainst = Scheme(
+              quantifiers=List(a.sym, ef.sym, opt1.sym, opt2.sym),
+              tconstrs = Nil,
+              econstrs = Nil,
+              base = Type.mkArrowWithEffect(
+                a=Type.mkArrowWithEffect(Type.Unit, ef, a, SourceLocation.Unknown),
+                // (ef - HandledEff) + (IO & opt1) + (NonDet & opt2)
+                p= Type.mkUnion(Type.mkDifference(ef, efHandle, SourceLocation.Unknown), optionalIO, optionalNonDet, SourceLocation.Unknown),
+                b=a,
+                loc=SourceLocation.Unknown
+              )
+          )
+          // Check if handler's scheme is valid
+          if(Scheme.lessThanEqual(schemeToCheckAgainst, handlerDef.spec.declaredScheme, root.traitEnv, root.eqEnv, Nil)){
+              Validation.Success(HandlerInfo(handlerSym, handlerDef, DefSymUse(handlerDef.sym, handlerDef.loc), efHandle, effSymbol, Nil))
+          } else {
+              Validation.Failure(EntryPointError.WrongSignatureForDefaultHandler(effSymbol, handlerSym.loc))
+          }
+        // The default handler is not in the companion module of an effect
```

Comment:
Move the None cast first, and put the comment UNDER the arrow.

// The default handler is NOT in the companion module of an effect. We issue an error.